fendermint/vm/interpreter/src/fvm/exec.rs

```diff
@@ -212,21 +219,47 @@ where
             PowerUpdates::default()
         };
 
-        // check for upgrades in the upgrade_scheduler
-        let chain_id = state.chain_id();
-        let block_height: u64 = state.block_height().try_into().unwrap();
-        if let Some(upgrade) = self.upgrade_scheduler.get(chain_id, block_height) {
-            // TODO: consider using an explicit tracing enum for upgrades
-            tracing::info!(?chain_id, height = block_height, "Executing an upgrade");
-
-            // there is an upgrade scheduled for this height, lets run the migration
-            let res = upgrade.execute(&mut state).context("upgrade failed")?;
-            if let Some(new_app_version) = res {
-                state.update_app_version(|app_version| {
-                    *app_version = new_app_version;
-                });
-
-                tracing::info!(app_version = state.app_version(), "upgraded app version");
+        if let Some(upgrade_info) = self.upgrade_schedule.get(state.block_height()) {
```

Comment:
I was thinking whether we should also add this check to the `ProposalInterpreter` implementations (which this interpreter does not implement), but I think you are right, this should be sufficient. If a validator proposes a block without knowing whether it will be able to execute it, it can still be finalized and cause all nodes to freeze there during delivery. 

If we live with this model, then it's also probably right to keep the execution of the upgrade in `end`, because the decision about what transactions to put into the block has been made the the current version of the software... but no, that doesn't make any difference, if the new version will have to execute it, it will have to be able to handle all the transactions that the previous one already put into the block. 

So _that_ could be prevented by adding the freeze already to the proposal handling: if for example we wanted to get rid of a certain transaction type, we could just say "we're no longer eligible to create the next block", rather than put it in, then consider it invalid. 

If we want to handle such cases by marking these transactions as invalid, then I reckon the upgrade should be done in `begin` after all, which signifies: once this block begins, it's the new rules, even if the transactions are old. Versus suggesting that "until the end, it's the old rules", which could leave the new version in a conundrum whether to be backwards compatible just for the sake of a single block.